x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,327 @@
+		go func(m sqs.Message) {
+			// launch goroutine to handle each message
+			defer wg.Done()
+
+			s3Infos, err := handleMessage(m)
+			if err != nil {
+				p.logger.Error(err.Error())
+			}
+
+			if err != nil {
+				p.logger.Error(err.Error())
+			}
+
+			// read from s3
+			events, err := readS3Object(svcS3, s3Infos)
+			if err != nil {
+				p.logger.Error(err.Error())
+			}
+
+			eventsTotal = append(eventsTotal, events...)
+			messagesReceiptHandles = append(messagesReceiptHandles, *m.ReceiptHandle)
+		}(messages[i])
+	}
+
+	wg.Wait()
+	return eventsTotal, messagesReceiptHandles, nil
+}
+
+// handle message
+func handleMessage(m sqs.Message) (s3Infos []s3Info, err error) {
+	msg := map[string]interface{}{}
+	err = json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		err = errors.Wrap(err, "json unmarshal sqs message body failed")
+		return
+	}
+
+	records := msg["Records"].([]interface{})
+	s3Info := s3Info{}
+	for _, record := range records {
+		recordMap := record.(map[string]interface{})
+		if recordMap["eventSource"] == "aws:s3" && recordMap["eventName"] == "ObjectCreated:Put" {
+			s3Record := recordMap["s3"].(map[string]interface{})
+			bucketInfo := s3Record["bucket"].(map[string]interface{})
+			objectInfo := s3Record["object"].(map[string]interface{})
+			s3Info.name = bucketInfo["name"].(string)
+			s3Info.key = objectInfo["key"].(string)
+			s3Infos = append(s3Infos, s3Info)
+		}
+	}
+	return
+}
+
+func readS3Object(svc s3iface.S3API, s3Infos []s3Info) ([]*beat.Event, error) {
+	var events []*beat.Event
+	for _, s3Info := range s3Infos {
+		s3GetObjectInput := &s3.GetObjectInput{
+			Bucket: awssdk.String(s3Info.name),
+			Key:    awssdk.String(s3Info.key),
+		}
+		objReq := svc.GetObjectRequest(s3GetObjectInput)
+
+		objResp, err := objReq.Send()
+		if err != nil {
+			return nil, errors.Wrap(err, "s3 get object request failed")
+		}
+
+		// TODO: check way to stream
+		buf := new(bytes.Buffer)
+		_, err = buf.ReadFrom(objResp.Body)
+		if err != nil {
+			return nil, errors.Wrap(err, "buf.ReadFrom failed")
+		}
+
+		s := buf.String() // Does a complete copy of the bytes in the buffer.
+		logLines := strings.Split(s, "\n")
+		for _, log := range logLines {
+			// create event
+			if log == "" {
+				continue
+			}
+			event := createEvent(log, s3Info)
+			events = append(events, event)
+		}
+	}
+	return events, nil
+}
+
+func deleteMessages(queueURL string, messagesReceiptHandles []string, svcSQS *sqs.SQS) error {
+	for _, receiptHandle := range messagesReceiptHandles {
+		deleteMessageInput := &sqs.DeleteMessageInput{
+			QueueUrl:      awssdk.String(queueURL),
+			ReceiptHandle: awssdk.String(receiptHandle),
+		}
+
+		req := svcSQS.DeleteMessageRequest(deleteMessageInput)
+		_, err := req.Send()
+		if err != nil {
+			return errors.Wrap(err, "DeleteMessageRequest failed")
+		}
+	}
+	return nil
+}
+
+func createEvent(log string, s3Info s3Info) *beat.Event {
+	f := common.MapStr{
+		"message": log,
+		"log": common.MapStr{
```

Comment:
`log` input also fills `log.offset`, I think it would make sense to add it too. It could also make sense to add the whole bucket/object path under `log.file.path`

Reply:
Is `log.offset` here means which line is the log message from in the log file?

Reply:
more than the line it tells you the byte where the line starts, more info about it here: https://github.com/elastic/beats/blob/master/filebeat/input/log/log.go#L33